cli/src/render.rs

```diff
@@ -0,0 +1,200 @@
+use std::collections::HashMap;
+use std::fmt::{Debug, Display};
+use std::io;
+use std::io::Write;
+
+use analyzer::diagnostic::ObservationTag;
+use ariadne::{Cache, Label, Report, ReportKind};
+use concolor::Stream;
+use yansi::{Color, Paint};
+
+use analyzer::relations::{ObjectId, SourceId};
+use context::source::Source;
+use parser::err::{ParseError, ParseErrorKind};
+use std::fmt::Write as FmtWrite;
+
+/// contains the sources bound to each [SourceId]
+pub struct SourcesCache<'a, S>
+where
+    S: Fn(SourceId) -> Source<'a>,
+{
+    /// supplier used to fetch a source bound to given [SourceId]
+    supplier: S,
+    names: HashMap<SourceId, String>,
+    sources: HashMap<String, ariadne::Source>,
+}
+
+impl<'a, S> SourcesCache<'a, S>
+where
+    S: Fn(SourceId) -> Source<'a>,
+{
+    pub fn new(supplier: S) -> Self {
+        Self {
+            supplier,
+            names: HashMap::new(),
+            sources: HashMap::new(),
+        }
+    }
+}
+
+impl<'a, S> Cache<SourceId> for SourcesCache<'a, S>
+where
+    S: Fn(SourceId) -> Source<'a>,
+{
+    fn fetch(&mut self, id: &SourceId) -> Result<&ariadne::Source, Box<dyn Debug + '_>> {
+        let pos = self.names.get(id);
+        match pos {
+            Some(name) => Ok(self.sources.get(name).expect("unable to find source")),
+            None => {
+                let source = (self.supplier)(*id);
+                self.names.insert(*id, source.name.to_string());
+                let src = self
+                    .sources
+                    .entry(source.name.to_string())
+                    .or_insert_with(|| ariadne::Source::from(source.source));
+                Ok(src)
+            }
+        }
```

Comment:
Map directly the `SourceId` type to the `ariadne::Source` type:
```rs
pub struct SourcesCache<'a, S>
where
    S: Fn(SourceId) -> Source<'a>,
{
    supplier: S,
    names: HashMap<SourceId, String>,
    sources: HashMap<SourceId, ariadne::Source>,
}
```
```suggestion
        let source = (self.supplier)(*id);
        self.names.insert(*id, source.name.to_string());
        Ok(self
            .sources
            .entry(*id)
            .or_insert_with(|| ariadne::Source::from(source.source)))
```

Reply:
No because this would make multiple source id bound to the same source, so duplicate the sources